src/collectors/loggly/loggly.py

```diff
@@ -0,0 +1,246 @@
+
+        def retrieve_results(metric_config):
+            def now_as_timefield():
+                if metric_config.timestamp_format is 'int':
+                    return time.mktime(datetime.datetime.now().timetuple())
+                else:
+                    return datetime.datetime.now().strftime(
+                        metric_config.timestamp_format)
+
+            def log_error_and_continue(code):
+                known_errors = {
+                    400: 'Bad Request',
+                    401: 'Unauthorized',
+                    403: 'Forbidden',
+                    404: 'Not Found',
+                    409: 'Duplicate',
+                    410: 'Gone',
+                    500: 'Internal Server Error',
+                    501: 'Not Implemented',
+                    503: 'Throttled',
+                    504: 'Gateway Timeout'
+                }
+                try:
+                    self.log.error('Loggly returned error %s: %s'
+                                   % (code, known_errors[code]))
+                except KeyError:
+                    self.log.error('Loggly returned unrecognized error code %s'
+                                   % code)
+                finally:
+                    return []
+
+            search_response = requests.get(
+                metric_config.search_URI,
+                auth=(self.config_bunch.username, self.config_bunch.password))
+            if search_response.status_code is 200:
+                events = requests.get(
+                    metric_config.events_URI % search_response
+                    .json()['rsid']['id'],
+                    auth=(self.config_bunch.username,
+                          self.config_bunch.password))
+
+                if events.status_code is 200:
+                    bare_metrics = [(metric_config.field,
+                                     event[metric_config.timestamp_field],
+                                     multi_getattr(bunch.bunchify(event).event,
+                                                   metric_config.field))
+                                    for event in events.json()['events']]
+                    if metric_config.count:
+                        bare_metrics.append(('count', now_as_timefield(),
+                                            events.json()['total_events']))
+                else:
+                    return log_error_and_continue(events.status_code)
+            else:
+                return log_error_and_continue(search_response.status_code)
+            return map(lambda bm: bunch.bunchify({
+                'field': bm[0],
+                'timestamp': parse_timestamp(metric_config, bm[1]),
+                'value': float(bm[2])}), bare_metrics)
+
+        def parallel_retrieval(results, metric, config):
+            results.update({metric: bunch.bunchify(
+                {'results': retrieve_results(config)})})
+
+        results = bunch.Bunch()
+        with concurrent.futures.ThreadPoolExecutor(
```

Comment:
This is a really nice feature to use from Python 3. I feel like other collectors should try and use it.